### DISCUSSION TITLE
Spotify Connect Devices

### DISCUSSION THREAD

**Discussion by gitisz** [librespot-org/librespot#996] [Feature Requests]
Not sure where to turn, so thought I would post here to reach some experts in the Spotify API + Connect skillset.

I have **librespot** installed and have registered three Spotify Sources (one for each member of the fam):

```
source = librespot:///librespot?name=MOM&devicename=Snapserver&bitrate=320&volume=100
source = librespot:///librespot?name=DAD&devicename=Snapserver&bitrate=320&volume=100
source = librespot:///librespot?name=SON&devicename=Snapserver&bitrate=320&volume=100

```

They all show up in our Spotify apps like our Android Phone App, or even Desktop Apps.  And even a bunch of others show up, like our NVidia Shield,  Firestick, and old Gramofons.

However, when I call https://api.spotify.com/v1/me/player/devices, depending on the Auth token for each one of my family, I get back only a couple devices, and definitely not all devices on the home network.

Curious... if the Spotify Desktop or Mobile App can get all the Devices on a home network, how can I?

This leads me to whether this capability would fall within the scope of **librespot**, and perhaps this Library could or already does support essentially what these other client apps already do.

Goal: If I could get back all these devices from the Spotify API, then my client app would allow my family to have a one-stop shop for grouping Snapcast clients, turning on network receivers, **and choosing playback device**. 

LMK if I am not making any sense, but thank you for taking time to read my inquiry!

**Comment by kingosticks** [comment]
Probably worth mentioning that those are snapcast sources specified in the post. 

Note that Spotify devices within the same LAN can see each other without needing to communicate via Spotify's server. That's possibly the reason for the discrepancy you are seeing. Although I always thought that librespot clients did publish themselves to Spotify's server and would show up in the results. I would guess only librespot servers that are currently logged in as that user will be present in the Spotify API results obtained using a given user's auth token. But I'm not sure, I've never tried that.

**Reply by gitisz** [reply]
All components, custom app, librespot, Snapcast, Mobile App, Desktop App, are all on the same network.  

+1 for calling out Snapcast, just didn't want to overload the question.

**Comment by michaelherger** [comment]
It is my understanding that the API would only return players you had connected to before. Devices on your network on the other hand use local announcements to find players. They don’t talk to a server to find the players, but they send out a broadcast message to all devices on the network. And those who understand the request will respond, making themselves known to that device or application.

**Comment by gitisz** [comment]
> Devices on your network on the other hand use local announcements to find players. 

That is how I understand it to work.  But somehow when a user selects the Device it calls Spotify's API servers to register it (unpublished API?) then the app proceeds to transfer playback to that source's Device ID (published API).  

The problem here I am calling out is for an app that any of us would like to develop, that integrates with with other capabilities, your/my app cannot get all the Devices the way the proprietary Spotify App can, and can only get just those Devices that have been recently selected by the active user from Spotify's App (very limiting).  

Also, their [TransferPlayback ](https://developer.spotify.com/documentation/web-api/reference/#/operations/transfer-a-users-playback)API call will only accept a DeviceId that has been registered for that user.

**Reply by shanemeagher** [reply]
> That is how I understand it to work. But somehow when a user selects the Device it calls Spotify's API servers to register it (unpublished API?) then the app proceeds to transfer playback to that source's Device ID (published API).

I think it is the other way around. A client device (say an Android app) finds a local connect device (say Librespot), connects to it and gives it an authentication blob. The connect device then logs into the Spotify server, from where the client app can tell it what to do.

You want a client that will discover a playback device (possible using Zeroconf), and then add it to the spotify account by supplying an authentication blob to it. Librespot is only a playback device at the moment, and I don't think there are any connect clients other than official Spotify ones (mobile apps, web player).

The Home Assistant project have the same issue as they can't add devices to an account, only use existing devices connected to the Spotify account.
https://github.com/home-assistant/core/issues/68066#issuecomment-1067319480

**Reply by kingosticks** [reply]
I think one of the related Go-based projects is a control client rather than a playback device. However, unsure if it supports registering playback devices with Spotify's server.

**Reply by shanemeagher** [reply]
Thinking of [SpotControl](https://github.com/badfortrains/spotcontrol)?

**Reply by kingosticks** [reply]
Yep!

**Reply by kingosticks** [reply]
I've not looked at it but I am assuming the difference is where librespot mostly ignores control commands from other devices, SpotControl processes them. I wanted to add this support to librespot (library) but never gotten around to it.

**Reply by gitisz** [reply]
This is all very helpful!  For the moment, I will take a look at [SpotControl](https://github.com/badfortrains/spotcontrol) and see how it works.  

But given this sounds like a feature librespot could/should have, does it make sense to turn this into an enhancement issue?  From the HomeAssistant (where I started from) issue, it sounds like ZeroConf is required, and such depends on being an approved partner.

**Reply by kingosticks** [reply]
> does it make sense to turn this into an enhancement issue

Maybe but don't be surprised if it's considered out of scope. Such a feature makes sense in a generic Spotify library but not in a library that serves to replace libspotify or a library focused on playback. Librespot's identify is a little muddy sometimes.

**Reply by roderickvd** [reply]
I'd be fine with getting it in. Sure, `librespot` has been focusing on doing its own playback a lot over the past year. But in `new-api` I'm again adding lots of stuff that `librespot-playback` has no use for. In doing so, aiming `librespot` to provide all functionality anyone could ask of a Spotify library. Except the open web API that any developer may access with his/her own library of choice.

**Comment by gitisz** [comment]
Hopefully these screenshots help show some differences:

Mom's Login Auth Token
![image](https://user-images.githubusercontent.com/4492953/166556577-74ba459b-b6b3-49d0-9686-15c5b9b62b4f.png)
Dad's Login Auth Token
![image](https://user-images.githubusercontent.com/4492953/166556610-421f2d53-abc0-4c55-b604-ed2999b85bd0.png)
All Device Options (logged in as Mom):
![image](https://user-images.githubusercontent.com/4492953/166556740-ffa8983e-3432-4acc-a078-a2626b51ff7a.png)

Mom or Dad, should be able to choose any device on a custom app, like they can from a Spotify App.
